# Seminar 5 - Answers

The data file <b>birthweight_smoking.xls</b> contains data for a random sample of babies born in Pennsylvania in 1989. 

The data include the baby's birth weight together with various characteristics of the mother, including whether she smoked during her pregnancy.

You can download the data from [here](https://www.princeton.edu/~mwatson/Stock-Watson_3u/Students/EE_Datasets/birthweight_smoking.xlsx).

A detailed description is given [here](https://www.princeton.edu/~mwatson/Stock-Watson_3u/Students/EE_Datasets/Birthweight_Smoking_Description.pdf). 

The dataset is used in the reference textbook Introduction to Econometrics, 4th edition 
(Stock and Watson).

In the exercises below we will investigate the relationship between birth weight and smoking during pregnancy.

Using this dataset, answer the following questions:

In [1]:
# Importing useful libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf
import scipy.stats as st
from tabulate import tabulate

In [4]:
# Loading the data set
df = pd.read_excel("Growth.xlsx")
df_xm = df[df.country_name!="Malta"]
df_xm

country_name    growth  oil       rgdp60  tradeshare  yearsschool  \
0           India  1.915168    0   765.999817    0.140502         1.45   
1       Argentina  0.617645    0  4462.001465    0.156623         4.99   
2           Japan  4.304759    0  2953.999512    0.157703         6.71   
3          Brazil  2.930097    0  1783.999878    0.160405         2.89   
4   United States  1.712265    0  9895.003906    0.160815         8.66   
..            ...       ...  ...          ...         ...          ...   
59    Netherlands  2.200577    0  6076.997070    0.834204         5.27   
60         Cyprus  5.384184    0  2037.000366    0.979355         4.29   
61       Malaysia  4.114544    0  1420.000244    1.105364         2.34   
62        Belgium  2.651335    0  5495.001953    1.115917         7.46   
63      Mauritius  3.024178    0  2861.999268    1.127937         2.44   

    rev_coups  assasinations  
0    0.133333       0.866667  
1    0.933333       1.933333  
2    0.000000       0.200000  
3    0.100000       0.100000  
4    0.000000       0.433333  
..        ...            ...  
59   0.000000       0.000000  
60   0.100000       0.166667  
61   0.033333       0.033333  
62   0.000000       0.000000  
63   0.000000       0.000000  

[64 rows x 8 columns]

<b> 1. Construct a table that shows the sample mean, standard deviation, and minimum and maximum values for the series Growth, TradeShare, YearsSchool, Oil, Rev_Coups, Assassinations, and RGDP60. Use the function <i>describe()</i></b>

In [5]:
df.describe()

growth   oil       rgdp60  tradeshare  yearsschool  rev_coups  \
count  65.000000  65.0    65.000000   65.000000    65.000000  65.000000   
mean    1.942715   0.0  3103.784649    0.564703     3.985077   0.167450   
std     1.897120   0.0  2512.656846    0.289270     2.542000   0.224680   
min    -2.811944   0.0   366.999939    0.140502     0.200000   0.000000   
25%     0.838156   0.0  1147.999756    0.393251     1.940000   0.000000   
50%     1.975147   0.0  2019.000244    0.543337     3.650000   0.066667   
75%     2.880327   0.0  5143.000977    0.681555     5.560000   0.266667   
max     7.156855   0.0  9895.003906    1.992616    10.070000   0.970370   

       assasinations  
count      65.000000  
mean        0.277564  
std         0.491528  
min         0.000000  
25%         0.000000  
50%         0.100000  
75%         0.233333  
max         2.466667

<b> 2. Run a regression of Growth on TradeShare, YearsSchool, Rev_Coups, Assasinations, and RGDP60. Is there a problem with the regression? why?

In [6]:
formula = 'growth ~ tradeshare + yearsschool + rev_coups + assasinations + rgdp60'

model = smf.ols(formula,df_xm).fit() 
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 growth   R-squared:                       0.291
Model:                            OLS   Adj. R-squared:                  0.230
Method:                 Least Squares   F-statistic:                     4.764
Date:                Mon, 04 Mar 2024   Prob (F-statistic):            0.00103
Time:                        14:16:25   Log-Likelihood:                -117.49
No. Observations:                  64   AIC:                             247.0
Df Residuals:                      58   BIC:                             259.9
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         0.6269      0.783      0.801

In order to avoid problems of numerical instability, we should change the units of the variable rgdp60

<b> 3. Rewrite the variable rgdp60 by changing its units to thousands of dollars. What is the value of the coefficient on Rev_Coups? Interpret the value of this coefficient. Is the coefficient significant? </b>

In [11]:
df_xm['rdgp60_000s'] = df_xm.rgdp60/1000

C:\Users\ecacicedo\AppData\Local\Temp\ipykernel_564\3184317573.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_xm['rdgp60_000s'] = df_xm.rgdp60/1000


In [12]:
formula = 'growth ~ tradeshare + yearsschool + rev_coups + assasinations + rdgp60_000s'

model = smf.ols(formula,df_xm).fit() 
print(model.summary()) 

                            OLS Regression Results                            
Dep. Variable:                 growth   R-squared:                       0.291
Model:                            OLS   Adj. R-squared:                  0.230
Method:                 Least Squares   F-statistic:                     4.764
Date:                Mon, 04 Mar 2024   Prob (F-statistic):            0.00103
Time:                        14:18:23   Log-Likelihood:                -117.49
No. Observations:                  64   AIC:                             247.0
Df Residuals:                      58   BIC:                             259.9
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         0.6269      0.783      0.801

The estimated regression equation is

$$ \hat{growth} = 0.62 + 1.34 \times tradeshare + 0.56 \times yearsschool - 2.15 \times rev\_coups + 0.32 \times assasinations -0.46 \times rgdp60 $$

$$ (0.78) \hspace{0.2in} (0.96) \hspace{1.1in} (0.14) \hspace{1.1in} (1.12) \hspace{0.9in} (0.49) \hspace{1.2in} (0.151) $$

The value of the coefficient on rev_coups is -2.15. 

This means that for any additional number of revolutions, insurrections and coup d’etats in the country in average from 1960 to 1995, the average growth percentage rate from 1960 to 1995 decreases by 2.15%. This makes sense as more instability should lead to a lower GDP growth.

The coefficient is not significant at 5% as its p-value is 5.9% (0.059).

<b> 4. Use the regression to predict the average annual growth rate for a country that has average values for all regressors. </b>

In [19]:
df_xm

country_name    growth  oil       rgdp60  tradeshare  yearsschool  \
0           India  1.915168    0   765.999817    0.140502         1.45   
1       Argentina  0.617645    0  4462.001465    0.156623         4.99   
2           Japan  4.304759    0  2953.999512    0.157703         6.71   
3          Brazil  2.930097    0  1783.999878    0.160405         2.89   
4   United States  1.712265    0  9895.003906    0.160815         8.66   
..            ...       ...  ...          ...         ...          ...   
59    Netherlands  2.200577    0  6076.997070    0.834204         5.27   
60         Cyprus  5.384184    0  2037.000366    0.979355         4.29   
61       Malaysia  4.114544    0  1420.000244    1.105364         2.34   
62        Belgium  2.651335    0  5495.001953    1.115917         7.46   
63      Mauritius  3.024178    0  2861.999268    1.127937         2.44   

    rev_coups  assasinations  rdgp60_000s  
0    0.133333       0.866667     0.766000  
1    0.933333       1.933333     4.462001  
2    0.000000       0.200000     2.954000  
3    0.100000       0.100000     1.784000  
4    0.000000       0.433333     9.895004  
..        ...            ...          ...  
59   0.000000       0.000000     6.076997  
60   0.100000       0.166667     2.037000  
61   0.033333       0.033333     1.420000  
62   0.000000       0.000000     5.495002  
63   0.000000       0.000000     2.861999  

[64 rows x 9 columns]

In [20]:
avg_growth = model.params.Intercept + model.params.tradeshare*np.mean(df_xm.iloc[:,4]) + model.params.yearsschool*np.mean(df_xm.iloc[:,5]) + model.params.rev_coups*np.mean(df_xm.iloc[:,6]) + model.params.assasinations*np.mean(df_xm.iloc[:,7]) + model.params.rdgp60_000s*np.mean(df_xm.iloc[:,8])
avg_growth

1.8691197348525754

<b> 5. Repeat 4., but now assume that the country’s value for TradeShare is one standard deviation above the mean. </b>

In [21]:
avg_growth2 = model.params.Intercept + model.params.tradeshare*(np.mean(df_xm.iloc[:,4]) + np.std(df_xm.iloc[:,4])) + model.params.yearsschool*np.mean(df_xm.iloc[:,5]) + model.params.rev_coups*np.mean(df_xm.iloc[:,6]) + model.params.assasinations*np.mean(df_xm.iloc[:,7]) + model.params.rdgp60_000s*np.mean(df_xm.iloc[:,8])
avg_growth2

2.1728713231729015

<b> 6. Why is Oil omitted from the regression? What would happen if it were included? </b>

In [22]:
formula2 = 'growth ~ tradeshare + yearsschool + rev_coups + assasinations + rgdp60 + oil'

model2 = smf.ols(formula2,df_xm).fit() 
print(model2.summary()) 

                            OLS Regression Results                            
Dep. Variable:                 growth   R-squared:                       0.291
Model:                            OLS   Adj. R-squared:                  0.230
Method:                 Least Squares   F-statistic:                     4.764
Date:                Mon, 04 Mar 2024   Prob (F-statistic):            0.00103
Time:                        14:23:35   Log-Likelihood:                -117.49
No. Observations:                  64   AIC:                             247.0
Df Residuals:                      58   BIC:                             259.9
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         0.6269      0.783      0.801

C:\ProgramData\Anaconda3\lib\site-packages\statsmodels\regression\linear_model.py:1918: RuntimeWarning: divide by zero encountered in double_scalars
  return np.sqrt(eigvals[0]/eigvals[-1])


The oil variable has all of its values equals to zero: no country had oil accounted for at least half of exports in 1960.

The result is that the determinant of the X matrix is zero (the matrix is singular) and therefore, the results of the OLS are unreliable.

## Exercise 2

The file <b>microsoft.xlsx</b> contains data on the daily share price of Microsoft from 2002 to 2022. It also contains the Excess return of the market and other four factors that can help explain the excess returns of Microsoft shares.

The factor and market data were obtained from Kenneth French's website [here](https://mba.tuck.dartmouth.edu/pages/faculty/ken.french/data_library.html) and they are part of the [Fama and French (1993)](https://www.sciencedirect.com/science/article/abs/pii/0304405X93900235) three factor model and [Fama and French (2013)](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2287202) five factor model.

[Here](https://mba.tuck.dartmouth.edu/pages/faculty/ken.french/data_library/f-f_5_factors_2x3.html#:~:text=The%20Fama%2FFrench%205%20factors,formed%20on%20size%20and%20investment.) is a brief description of the five factor model.

The Column RF corresponds to the risk-free rate and the column Mkt-Rf is the excess return of the markets.

In [24]:
data = pd.read_excel("microsoft.xlsx")
data

Date  Mkt_RF   SMB   HML   RMW   CMA     RF   microsoft
0    2002-01-22   -0.81 -0.19  0.03  1.14  1.11  0.007   20.108234
1    2002-01-23    0.92  0.62 -0.16 -0.38 -0.54  0.007   19.883644
2    2002-01-24    0.40 -0.18  0.29 -0.68 -0.09  0.007   20.151907
3    2002-01-25    0.09 -0.15  0.30  0.53  0.42  0.007   19.902351
4    2002-01-28    0.05  0.21  0.66  0.23  0.05  0.007   19.908590
...         ...     ...   ...   ...   ...   ...    ...         ...
5269 2022-12-23    0.51 -0.34  1.15  0.86  0.46  0.016  238.133545
5270 2022-12-27   -0.51 -0.42  1.42  1.13  1.21  0.016  236.367981
5271 2022-12-28   -1.23 -0.30 -0.29 -0.96 -0.03  0.016  233.944031
5272 2022-12-29    1.87  1.04 -1.07 -1.01 -0.82  0.016  240.407837
5273 2022-12-30   -0.22  0.12 -0.03 -0.53  0.02  0.016  239.220825

[5274 rows x 8 columns]

<b> 1. Create a new column to the dataset with the excess return of Microsoft. The excess return is

$$  R_{Microsoft} - R_{risk-free} $$

</b>

In [25]:
data['ermicrosoft'] =  100*np.log(data.microsoft/data.microsoft.shift(1)) - data.RF

data2 = data.dropna()

<b> 2. Run the Capital Asset Pricing Model for Microsoft Returns whose population regression is

$$ R_{Microsoft} - R_{risk-free} = \beta_{0} + \beta_{1} \times (R_{Market} - R_{risk-free}) $$

What is the coefficient for the excess return of the market? What is the meaning? Is it significant? Write down the estimated regression line. </b>

In [26]:
formula2 = 'ermicrosoft ~ Mkt_RF'

model2 = smf.ols(formula2,data2).fit() 
model2.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            ermicrosoft   R-squared:                       0.546
Model:                            OLS   Adj. R-squared:                  0.545
Method:                 Least Squares   F-statistic:                     6328.
Date:                Mon, 04 Mar 2024   Prob (F-statistic):               0.00
Time:                        15:55:11   Log-Likelihood:                -8415.4
No. Observations:                5273   AIC:                         1.683e+04
Df Residuals:                    5271   BIC:                         1.685e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0052      0.016      0.316      0.752      -0.027       0.037
Mkt_RF         1.0489      0.013     79.547      0.000       1.023       1.075
==============================================================================
Omnibus:                     1195.881   Durbin-Watson:                   2.037
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            33512.616
Skew:                          -0.443   Prob(JB):                         0.00
Kurtosis:                      15.319   Cond. No.                         1.25
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

<b>
3. Run the three factor model whose population regression line is

$$ R_{Microsoft} - R_{risk-free} = \beta_{0} + \beta_{1} \times (R_{Market} - R_{risk-free}) + \beta_{2} \times SMB + \beta_{3} \times HML$$

Explain the meaning of the coefficients. Are they significant? Was there an improvement in fit? Write down the estimated regression line.
</b>

In [27]:
formula3 = 'ermicrosoft ~ Mkt_RF + SMB + HML'

model3 = smf.ols(formula3,data2).fit() 
model3.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            ermicrosoft   R-squared:                       0.599
Model:                            OLS   Adj. R-squared:                  0.599
Method:                 Least Squares   F-statistic:                     2626.
Date:                Mon, 04 Mar 2024   Prob (F-statistic):               0.00
Time:                        15:55:33   Log-Likelihood:                -8084.3
No. Observations:                5273   AIC:                         1.618e+04
Df Residuals:                    5269   BIC:                         1.620e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0068      0.015      0.438      0.661      -0.024       0.037
Mkt_RF         1.1227      0.013     88.324      0.000       1.098       1.148
SMB           -0.3902      0.026    -15.062      0.000      -0.441      -0.339
HML           -0.3965      0.021    -19.210      0.000      -0.437      -0.356
==============================================================================
Omnibus:                     1391.226   Durbin-Watson:                   2.060
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            51986.479
Skew:                          -0.553   Prob(JB):                         0.00
Kurtosis:                      18.343   Cond. No.                         2.17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The coefficients for Small Minus Big and High Minus Low appear to be also signficant since all p-values are below 5%.

The coefficient of SMB is -0.39 which means that an increase in 1% in the gap of returns between small and big shares will decrease the return of Microsoft shares by 0.39%, for the same level of market excess return and the gap between high and low shares.

The coefficient of HML is -0.40 which means that an increase in 1% in the gap of returns between shigh and low shares will decrease the return of Microsoft shares by 0.40%.

The fit of the regression line improved from an Adjusted R sqaured of 0.545 to 0.599, which means that the three factors are able to explain approximately 60% of the variability in returns for Microsoft.

<b>
4. Run the five factor model whose population regression line is

$$ R_{Microsoft} - R_{risk-free} = \beta_{0} + \beta_{1} \times (R_{Market} - R_{risk-free}) + \beta_{2} \times SMB + \beta_{3} \times HML + \beta_{4} \times RMW + \beta_{5} \times CMA$$

Explain the meaning of the coefficients. Are they significant? Was there an improvement in fit? Write down the estimated regression line.
</b>

In [28]:
formula4 = 'ermicrosoft ~ Mkt_RF + SMB + HML + RMW + CMA'

model4 = smf.ols(formula4,data2).fit() 
model4.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            ermicrosoft   R-squared:                       0.599
Model:                            OLS   Adj. R-squared:                  0.599
Method:                 Least Squares   F-statistic:                     1576.
Date:                Mon, 04 Mar 2024   Prob (F-statistic):               0.00
Time:                        15:56:25   Log-Likelihood:                -8083.2
No. Observations:                5273   AIC:                         1.618e+04
Df Residuals:                    5267   BIC:                         1.622e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0064      0.015      0.412      0.680      -0.024       0.037
Mkt_RF         1.1243      0.014     82.295      0.000       1.097       1.151
SMB           -0.3833      0.027    -14.444      0.000      -0.435      -0.331
HML           -0.3896      0.023    -16.678      0.000      -0.435      -0.344
RMW            0.0412      0.035      1.173      0.241      -0.028       0.110
CMA           -0.0415      0.047     -0.874      0.382      -0.134       0.052
==============================================================================
Omnibus:                     1400.507   Durbin-Watson:                   2.060
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            52361.192
Skew:                          -0.563   Prob(JB):                         0.00
Kurtosis:                      18.397   Cond. No.                         4.05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The new added factors RMW and CMA are not significant, since their p-values are above 5%. Moreover, the fit of the regression has not improved. The adjusted R squared plateaued at 0.599.